From 9f34a9ffe5b94aee93baa5762719546e0bebc28d Mon Sep 17 00:00:00 2001
From: Kai Lueke <kailuke@microsoft.com>
Date: Fri, 15 Mar 2024 11:49:50 +0100
Subject: [PATCH] x86/boot: Add space for dm-verity hash storage

Add space in the boot sector for storing dm-verity hash data, which is
needed for Flatcar's verified boot implementation. This preserves the
functionality that was available in older kernels but adapts it to the
new boot sector layout in 6.16.
---
 arch/x86/boot/header.S | 43 +++++++++++++++++++++++++++++++++++++++++++
 arch/x86/boot/setup.ld |  7 ++++---
 2 files changed, 47 insertions(+), 3 deletions(-)

--- a/arch/x86/boot/header.S
+++ b/arch/x86/boot/header.S
@@ -41,6 +41,49 @@ SYSSEG		= 0x1000		/* historical load address >> 4 */

 	.code16
 	.section ".bstext", "ax"
+
+	.global bootsect_start
+bootsect_start:
+	# Normalize the start address
+	ljmp	$BOOTSEG, $start2
+
+start2:
+	movw	%cs, %ax
+	movw	%ax, %ds
+	movw	%ax, %es
+	movw	%ax, %ss
+	xorw	%sp, %sp
+	sti
+	cld
+
+	movw	$bugger_off_msg, %si
+
+msg_loop:
+	lodsb
+	andb	%al, %al
+	jz	bs_die
+	movb	$0xe, %ah
+	movw	$7, %bx
+	int	$0x10
+	jmp	msg_loop
+
+bs_die:
+	# Allow the user to press a key, then reboot
+	xorw	%ax, %ax
+	int	$0x16
+	int	$0x19
+
+	# int 0x19 should never return.  In case it does anyway,
+	# invoke the BIOS reset code...
+	ljmp	$0xf000,$0xfff0
+
+	.section ".bsdata", "a"
+bugger_off_msg:
+	.ascii	"Use a boot loader.\r\n"
+	.ascii	"\n"
+	.ascii	"Remove disk and press any key to reboot...\r\n"
+	.byte	0
+
 #ifdef CONFIG_EFI_STUB
 	# "MZ", MS-DOS header
 	.word	IMAGE_DOS_SIGNATURE
--- a/arch/x86/boot/setup.ld
+++ b/arch/x86/boot/setup.ld
@@ -11,10 +11,11 @@ ENTRY(_start)
 SECTIONS
 {
 	. = 0;
-	.bstext : {
-		*(.bstext)
-		. = 495;
-	} =0xffffffff
+	.bstext		: { *(.bstext) }
+	.bsdata		: { *(.bsdata) }
+
+	. = 495;
+	. = ALIGN(1);

 	.header		: { *(.header) }
 	.entrytext	: { *(.entrytext) }
